File: commands/dd/levelinfo.ts
import { ChatInputCommandInteraction, EmbedBuilder, SlashCommandSubcommandBuilder } from 'discord.js';
import Bot from '../../bot/bot';
import SlashCommand from '../../bot/slashcommand';
import levelData from '../../database/dungeon_defenders/leveldata.json';

export default class LevelInfo extends SlashCommand {
  constructor(client: Bot) {
    const data = new SlashCommandSubcommandBuilder()
      .setName('levelinfo')
      .setDescription('View level requirements for different items');
    super(data, client, { belongsTo: 'dd' });
  }

  async execute(interaction: ChatInputCommandInteraction): Promise<void> {
    const padder = (name: string, weapon: string, armor: string, pet: string) =>
      `${name.padEnd(14)}${weapon.padEnd(9)}${armor.padEnd(9)}${pet}`;

    const desc = [padder('Name/Level', 'Weapons', 'Armor', 'Pets')];

    for (const quality of levelData) {
      desc.push(padder(quality.name, quality.levels.weapons, quality.levels.armor, quality.levels.pets));
    }

    const embed = new EmbedBuilder()
      .setColor(this.client.config.colors.BLUE)
      .setTitle('Level Info')
      .setDescription(`\`\`\`${desc.join('\n')}\`\`\``);
    interaction.reply({ embeds: [embed] });
  }
}

File: src/commands/dd/levelinfo.rs
use serde::Deserialize;
use serenity::all::{CommandInteraction, CommandOptionType};
use serenity::async_trait;
use serenity::builder::{CreateCommandOption, CreateEmbed};
use serenity::client::Context;

use crate::color::Colors;
use crate::commands::SlashSubCommand;
use crate::database::Database;
use crate::interaction::BetterResponse;
use crate::Result;

#[derive(Deserialize)]
struct Quality {
  name: String,
  levels: Levels
}

#[derive(Deserialize)]
struct Levels {
  armor: i32,
  pets: i32,
  weapons: i32
}

#[derive(Default)]
pub struct LevelInfo;

#[async_trait]
impl SlashSubCommand for LevelInfo {
  fn register(&self) -> CreateCommandOption {
    CreateCommandOption::new(
      CommandOptionType::SubCommand,
      "levelinfo",
      "View level requirements for different items"
    )
  }

  async fn execute(&self, ctx: &Context, interaction: &CommandInteraction, _: &Database) -> Result<()> {
    let file = include_str!("../../../assets/dungeon_defenders/leveldata.json");
    let qualities: Vec<Quality> = serde_json::from_str(file).expect("JSON was not well formatted.");

    let desc = [padder("Name", "Weapons", "Armor", "Pets"), "".into()]
      .into_iter()
      .chain(qualities.iter().map(|quality| {
        padder(
          &quality.name,
          quality.levels.weapons,
          quality.levels.armor,
          quality.levels.pets
        )
      }))
      .collect::<Vec<_>>();

    let embed = CreateEmbed::new()
      .color(Colors::Blue)
      .title("Level Requirements")
      .description(format!("```{}```", desc.join("\n")));

    interaction.reply_embed(ctx, embed).await?;
    Ok(())
  }
}

fn padder<N, T>(name: N, weapon: T, armor: T, pet: T) -> String
where
  N: ToString,
  T: ToString
{
  format!(
    "{: <14}{: <9}{: <9}{: <5}",
    name.to_string(),
    weapon.to_string(),
    armor.to_string(),
    pet.to_string()
  )
}
